### PR TITLE
feat: add customize content type parsers for api plugin

### PR DESCRIPTION
**Pull request by Josh-Walker-GM**
Alterations to #10449

### CODE REVIEW

File: docs/docs/docker.md [hunk 1]
```diff
@@ -613,19 +608,38 @@ For example, to support image file uploads you'd tell Fastify to allow `/^image\
 ```ts title="api/src/server.ts"
 const server = await createServer({
   logger,
-})
-
-server.addContentTypeParser(/^image\/.*/, (req, payload, done) => {
-  payload.on('end', () => {
-    done()
-  })
+  configureApiServer(server){
```
Comment (Josh-Walker-GM) [inline docs/docs/docker.md:611]: Need to edit this so it's not `server` oops

File: docs/docs/docker.md [hunk 2]
```diff
@@ -577,21 +570,23 @@ Then register them with the appropriate config:
 ```ts title="api/src/server.ts"
 const server = await createServer({
   logger,
+  async configureApiServer(server) {
```
Comment (Josh-Walker-GM) [inline docs/docs/docker.md:573]: Again this parameter can't be called `server` in this code right here

File: packages/api-server/src/plugins/api.ts [hunk 3]
```diff
@@ -54,6 +56,11 @@ export async function redwoodFastifyAPI(
     }
   }
 
+  // Run users custom server configuration function
```
Comment (dthyresson) [inline packages/api-server/src/plugins/api.ts:59]: Note @Josh-Walker-GM and I found that the order where custom server config was added matters - if after loading functions, then plugins like fastify complress did not work properly.

**Review by dthyresson — CHANGES_REQUESTED** [review]
@Josh-Walker-GM I confirmed I could compress api requests:

image


But, graphql requests are not.

image


Can we add that -- or should we always compress both for api and graphql servers?

We'd still need this in createServer if the user decide to mount and server more that just the api and graphql, but we do want to let people compress graphql responses.

Also - is this a breaking change? If people use register, they will have to reconfigure their server.

**Comment by dthyresson** [comment]
@Josh-Walker-GM Should we add:

```
async function main() {
  const server = await createServer({
    logger,
    async configureApiServer(s) {
      await s.register(import('@fastify/compress'), {
        global: true,
        threshold: 1024,
        encodings: ['deflate', 'gzip'],
      })
    },
    async configureGraphQLServer(s) {
      await s.register(import('@fastify/compress'), {
        global: true,
        threshold: 1024,
        encodings: ['deflate', 'gzip'],
      })
    },
``` ?

In case want graphql function to handle differently?

**Review by dthyresson — APPROVED** [review]

**Comment by dthyresson** [comment]
@Josh-Walker-GM sounds good I’ll make an issue for GraphQL compression as well

File: docs/docs/docker.md [hunk 4]
```diff
@@ -613,19 +608,38 @@ For example, to support image file uploads you'd tell Fastify to allow `/^image\
...
 const server = await createServer({
   logger,
-})
-
-server.addContentTypeParser(/^image\/.*/, (req, payload, done) => {
-  payload.on('end', () => {
-    done()
-  })
+  configureApiServer(s){
+    s.addContentTypeParser(/^image\/.*/, (req, payload, done) => {
```
Comment (Tobbe) [inline docs/docs/docker.md:612]:
```suggestion
    server.addContentTypeParser(/^image\/.*/, (_req, payload, done) => {
```
`req` is not used

File: docs/docs/docker.md [hunk 5]
```diff
@@ -577,21 +570,23 @@ Then register them with the appropriate config:
 ```ts title="api/src/server.ts"
 const server = await createServer({
   logger,
+  async configureApiServer(s) {
```
[the remark this answers is not in the document]
Comment (Tobbe) [reply docs/docs/docker.md:573]: Interesting. I'd assume it would shadow the outer `server` variable 🤔

**Pull request merged**